시험 권한인증 페이지 / 권한 인증/ 응시페이지 / 제출 기능 구현

kihong1602 opened the pull request:
# Abtract
- 시험 응시권한 인증 페이지 구현
- 권한 인증 기능 구현
- 시험 응시 페이지 구현
- 시험 답안 제출 기능 구현

# Detail of Changes
## 시험 응시권한 인증페이지 구현
- 이메일로 권한인증을 위한 링크를 보낸다고 가정하였습니다.
- 쿼리스트링의 수험번호를 이용해 view로 해당 시험의 채용공고 제목 등의 데이터를 출력하고,
   - 추가로 UUID를 생성해 Cookie의 Value, session Attribute의 Key로, 수험번호를 Session Attribute의 value로 넣어 권한 인증 프로세스에서 사용할 수 있도록 하였습니다.
- input태그를 넣어 수험번호를 입력할 수 있게 하였습니다.
<details><summary>추가</summary>
<p>
추후에는 쿼리스트링값을 복잡하게하여 URL을 보고 유추해석할 수 없게 처리할 예정입니다.
</p>
</details> 

>  /exam/auth/{exam_id}?{queryString}  `GET`
- Controller: ExamAuthController `doGet()`
파라미터의 수험번호를 통해 시험데이터를 조회해 view로 forward 합니다.
- DTO: RecruitInfoDTO
해당 시험의 채용공고 데이터를 담을 객체입니다.
- Service: ExamService `getRecruitContent()`
apt_id가 null인지 검증하고, 검증 결과에 따라 채용공고 데이터 또는 null을 반환합니다.
- DAO: ExamDAO
파라미터로 받은 apt_id로 recruitment 테이블과의 join연산을 통해 채용공고 데이터를 조회합니다.
- View: `exam-auth.jsp`
전달받은 채용공고 데이터를 페이지에 출력합니다.



## 응시권한 인증 기능
 - 수험번호를 입력하고 제출하기 버튼을 클릭하면 자바스크립트로 비동기 통신을 합니다.
- 페이지에서 입력받은 수험번호를 Cookie -> Session을 통해 조회한 수험번호와 대조합니다.
- 수험번호가 일치한다면 available을, 불일치한다면 unavailable을 JSON 데이터로 만들어 응답합니다.
- 응답받은 data를 통해 avaiable이라면 `/exam/{exam_id'로 GET요청을 보냅니다.
   - unavailable이라면 alert를 통해 '수험번호가 잘못되었습니다.'를 출력합니다.
   - 만약, examAuth 쿠키가 없는, url만 사용해 들어온 상황이라면, authFail을 응답으로 보내고, '인증정보가 없습니다.' alert를 출력하고 이전페이지로 이동합니다.

>  /exam/auth/{exam_id}?{queryString} `POST`
- Controller: ExamAuthController `doPost()`
입력받은 수험번호와 session의 수험번호를 대조합니다.
- CookieManager: `getSessionValue()` 메서드를 호출해 Cookie를 통해 session value에 담겨있는 apt_id를 조회합니다.

## 시험 응시페이지 구현
view에서는 전달받은 데이터를 List형식으로 배치합니다.

> /exam/{exam_id) `GET`
- Controller: ExaminationController `doGet()`
URL에서 exam_id를 추출한 다음, 해당 값으로 시험문제 데이터를 조회합니다.
- DTO: ExaminationDTO
`String question` 과 `Integer answer` 로 이루어진 시험데이터를 담는 객체입니다.
- Service: ExamService `getExamination()`
exam_id를 통해 recruitment_id를 조회하고, 조회한 채용공고ID를 통해 시험문제 데이터를 조회하는 ExamDAO를 호출합니다.
    - 현재 question과 answer 테이블이 없는관계로, 아래와 같이 레코드를 삽입하였기에 설명과 같은 프로세스로 진행하였습니다.

id | question | answer | create_date | modified_date | recruitment_id
---|-------|---------|-----------|-------------|------------|
1 | 1번문제입니다 | 1(정답) | 생성일자 | 수정일자 | 1|
2 | 2번문제입니다 | 2(정답) | 생성일자 | 수정일자 | 1|
3| 3번문제입니다 | 3(정답) | 생성일자 | 수정일자 | 1|
4 | 4번문제입니다 | 4(정답) | 생성일자 | 수정일자 | 1|
5 | 5번문제입니다 | 5(정답) | 생성일자 | 수정일자 | 1|

- DAO: ExamDAO `getRecruitId()` `getExamination()`
수험번호를 통해 채용공고 ID를 조회하고, 조회한 채용공고ID로 시험문제 데이터를 조회한 후 List 타입으로 반환합니다.

- View: `examination.jsp`
서버에서 전달받은 List<ExaminationDTO>를 jstl의 forEach를 통해 페이지에 출력합니다.

## 답안 제출 기능

> 현재 답안을 저장할 테이블이 없는 관계로, DAO까지 값이 전달되는 것만 확인하였습니다. 추후 테이블 재설계 및 리팩토링 시 수정 예정입니다.
- 제출하기 버튼 클릭시, 자바스크립트에서 fetch 비동기 통신이 시작됩니다.
- 전달받은 JSON 데이터를 객체에 담고, Service -> DAO로 전달해 DB에 저장합니다.

> /exam/{exam_id} `POST`
- JavaScript: examination.js `submit()` `submitExamination()`
    - jsp페이지에서 answer{i}형태로 p 태그의 id를 지정하여, 해당 id의 textContent를 조회해 "answer":  "answerData" 형태로 배열을 생성하였습니다. 
    - 해당 JSON형태의 배열을 `submitExamination()`에 파라미터로 사용하여, apt_id와 함꼐 json 타입으로 만들어 JSON 데이터로 변환하였습니다.
    - URL에서 apt_id, exam_id를 추출합니다.
    - 비동기 통신을 통해 받은 응답을 분기로 지정해 프로세스를 마무리합니다.
- Controller: ExaminationController 'doPost()'
입력된 JSON 데이터를 전달받아 Gson으로 객체화 합니다. JSON데이터의 구조는 다음과 같습니다.
```JSON
{
   "aptId": "apt_id",
   "answers":  [ 
        {"answer": "1"},
        {"answer": "2"},
        {"answer": "3"},
        {"answer": "4"},
        {"answer": "5"} 
    ]
}
```
- DTO: AnswerData, Answer
계층형 JSON 데이터를 담을 객체입니다. 구조는 다음과 같습니다.
```Java
import java.util.List;

    public class AnswerData {

       private String aptId;
       private List<Answer> answers;

    }

     public class Answer {

       private Integer answer;
     
    }
```
- Service: ExamService `saveExamination()`
입력된 답안을 DB에 저장하기 전, apt_id와 answers를 분리하는 작업을 실행합니다.
DAO에서 반환받은 값을 확인 한 후, Controller로 결과를 반환합니다.
현재는 DAO로 AnswerData 객체를 전달하는 프로세스만 수행 중 입니다.


- DAO: ExamDAO `saveExamination()`
ExamService에서 파라미터로 들어온 Answers 객체를 DB에 저장합니다.
저장후 결과를 ExamService로 반환합니다.


# To Others
다급하게 생각해 설계를 짜임새있게 하지않고 진행하여 중간 중간 엎고 다시 코드짜는 일이 여러번 발생하였습니다..
빨리 코드작성하려는 마음을 버리고 설계부터 제대로 해야겠다는 교훈을 얻었습니다..
<details><summary>주석</summary>
<p>
DB 테이블의 재설계를 해야할 것 같습니다. 저 너무 헷갈려요..
</p>
</details>

kihong1602 committed:
Chore: 개인폴더 및 프로젝트 생성

kihong1602 committed:
Merge branch 'main' of https://github.com/Team-RecruTe/RecruTe-Service

hossi97 committed:
Docs: PR 템플릿 디렉토리 변경

PR 템플릿 디렉토리 변경

kihong1602 committed:
Chore: 의존성추가, gitignore 설정 추가

의존성 : jstl,Gson,Mybatis,MySql 추가
gitignore.io에서 windows,intellij, java, maven 추가

hossi97 committed:
Chore: 프로젝트에 필요한 설정 적용

1. 의존성 추가
- mysql
- mybatis
- jstl
- gson

2. 코드 컨벤션 적용
- intellij-java-google-style.xml 추가 및 설정

3. gitginore 적용
- gitignore.io에서 java, mac, intellij, maven 추가
- Github 저장소에 gitignore 재적용

kihong1602 committed:
Chore: 의존성추가, gitignore 설정 추가, mybatis config 설정

의존성: jstl,Gson,Mybatis,MySql 추가
gitignore.io에서 windows,intellij, java, maven 추가
mybatis: config.xml 설정

2JISUN committed:
Chore : 프로젝트 설정, 의존성 추가

1. eclipse 코드 컨벤션
- eclipse-java-google-style.xml 추가

2. gitignore 추가 
- 자바, 메이븐, 이클립스, 윈도우 추가

3. 의존성 추가
- mysql
- mybatis
- jstl
- gson

jujuri committed:
Chore: 의존성 추가, 코드컨벤션, gitignore

1. 작업을 도와줄 mybatis, jstl, mysql, gson 을 추가하였다
2. 공동 작업에 용이하게 가독성을 높여주도록 코드컨벤션을 설정하였다
3. git에 업데이트 되지 않아도 되는 파일을 알려주는 코드를 gitignore파일에 넣었다.

kihong1602 committed:
Chore: 코드 컨벤션, 의존성 추가, gitignore 설정, mybatis config 설정

1. intellj-java-google-style.xml을 추가해 코드컨벤션 적용

2. 프로젝트에 필요한 의존성 추가
- mybatis: 3.5.13
- mysql-connector-j: 8.0.33
- mysql-connector-java: 8.0.33
- gson: 2.10.1
- jstl-api: 3.0.0
- jstl: 3.0.1

3. gitignore.io에서 windows, intellij, java, maven 으로 설정해 기존 프로젝트 .gitignore 변경
4. mybatis 연동을 위한 mybatis-config.xml 설정

kihong1602 committed:
Chore: front format 구현, saveaction 설정

layout: header, footer, main 파트 구현
saveactions 설정 후 auto formatting 설정

kihong1602 committed:
Chore: 회원가입 페이지 기능 구현
signup-process.jsp: 회원가입 form 구현
SignUpController: GET 요청시 signup-process.jsp로 이동

Issue: #57

kihong1602 committed:
Chore: css 추가

header.jsp에 layout.css를 link

kihong1602 committed:
Feat: 회원가입기능 구현, Mybatis-config 수정

- 회원가입 페이지, 회원가입 완료 페이지
- controller로 회원가입 정보 입력
- MemberDTO, MemberDAO, Service 구현

Issue: #57, #58

kihong1602 committed:
Chore: ViewResolver 클래스 생성, WEB-INF 디렉토리 재구성

- ViewResolver 클래스로 경로 생성
- 파트 구분을 명확하게 하기위한 WEB-INF 디렉토리를 재구성하였음.

kihong1602 committed:
Chore: MemberDTO 속성 타입 변경

birth의 Type을 String -> LocalDate로 변경.

kihong1602 committed:
Feat: 로그인 페이지 구현, 로그인 기능 구현

header: 로그인 form 페이지 링크
signin-process: 로그인 form 페이지
signin: 로그인 완료 페이지

로그인 기능
- javascript: logininvalid(): input value 검증
- SignInController
- member.xml: member_id,password 검증

Issue: #59, #60

kihong1602 committed:
Feat: ID 중복체크, password Confirm

- ID 중복체크: js의 fetch를 사용해 json을 이용한 비동기 통신으로 중복체크 기능 추가
- password confirm : js로 password와 password_confirm의 value를 비교해 유효성 검사

Issue: #57

kihong1602 committed:
Chore: js파일 코드 분리

- 회원가입 관련 js코드와 로그인 관련 js코드를 function.js에서 각각 signUp.js와 signIn.js에 분리해서 작성
- 이에 따른 script태그 위치 변경

kihong1602 committed:
Chore: 이메일 유효성검사 코드 수정

- 코드이동으로 인한 value의 오타 수정

kihong1602 committed:
Chore: 로그인 기능 비동기처리로 코드 변경,IdCheckVO 이름 변경, 인증쿠키 생성

- LoginController: 비동기처리로 로그인 기능을 수행할 controller(추후 SignInController와 병합 예정)
- IdCheckVO: IdCheckDTO로 이름 변경
- Authenticater: 로그인유지 기능을 위한 인증쿠키를 생성하는 기능을 구현

kihong1602 committed:
Chore: 로그인기능 코드 변경

- 로그인기능을 fetch로 비동기 통신을 통해 유효성 검증 및 회원정보 대조
- 구조 변경으로 인한 SigninController,signin.jsp 삭제

kihong1602 committed:
Chore: 로그인 관련 파일 이름 변경

- SignInController와 signin.jsp의 삭제로 기존 LoginController의 이름 변경
- mapping된 url 변경
- 기존 SignInController에서 사용하던 MemberService의 method 삭제

kihong1602 committed:
Feat: 로그아웃 기능 추가

- SignOutController: 인증쿠키가 있는지 확인한 후, 보유하고 있다면 쿠키를 삭제하고, index 페이지로 redirect하는 기능

Issue: #61

kihong1602 committed:
Chore: 코드 구조 개선

- 코드 가독성 향상을 위한 코드 구조 개선

kihong1602 committed:
Chore: JSON parsing 클래스 분리

- request body로 들어오는 JSON 을 파싱하는 클래스 생성

kihong1602 committed:
Chore: 회원가입 입력값 유효성 검증 정규식 추가

- 회원가입 시 입력받는 ID,Password,Name,Email,PhoneNumber에 대해 유효성 검사를 하기 위한 정규식 추가

kihong1602 committed:
Chore: SignInController 코드수정

- commitID 89242c2e의 JsonUtil 클래스의 추가로 인한 SignInController의 JSON parsing관련 코드 수정

kihong1602 committed:
Chore: 불필요한 Default값 삭제

- jsp 내 불필요한 default value 삭제
- e.printStackTrace() 를 Logger를 사용해 log를 출력하도록 변경
- controller 내부 static 필드 변수들을 final static 으로 수정

kihong1602 committed:
Chore: 회원가입 input value 검증

- 회원가입시 input value를 체크하여, 올바른 입력이 들어오는지 체크하게 기능 수정

kihong1602 committed:
Merge branch 'main' into main

kihong1602 committed:
Merge branch 'main' of https://github.com/kihong1602/kihong1602

kihong1602 committed:
Chore: 채용공고, 지원 기능 설정

- 채용공고, 채용 지원 기능 구조 설정

kihong1602 committed:
Chore: 채용공고, 지원 기능 설정 구성

- 채용공고페이지 구성
- 채용공고 지원 구성

kihong1602 committed:
Merge branch 'feature/recruit' of https://github.com/kihong1602/kihong1602 into feature/recruit

kihong1602 committed:
Chore: ViewResolver viewPath() 코드 수정, apply.jsp 삭제

- view 경로 설정 에러발생으로
"WEB-INF/"+viewName+".jsp"에서
"/WEB-INF/"+viewName+".jsp"로 변경
- 불필요한 apply.jsp 삭제

kihong1602 committed:
Chore: 회원가입 유효성검증 분기문 조건 누락 수정

- 유효성검증 invalid() 분기문 수정

kihong1602 committed:
Chore: DTO 구조 작성

- ApplyDTO: 채용 지원시 applicant 테이블에 입력할 데이터
- CompanyDTO: 회사정보 조회 DTO
- DetailDTO: recruitment_id를 이용해 채용공고 상세내용, 채용 정보, 회사정보를 조회 후 Response할  DTO
- RctDetDTO: 채용공고 상세내용 조회 DTO
- RecruitDTO: 채용공고 조회하기위한 DTO
- CompanyDTO,RctDetDTO : 미사용시 삭제예정

kihong1602 committed:
Feat: 채용공고 페이지열람 기능 구현

- rct-detail.jsp: Get 요청시 상세내용 페이지 구성
- RecruitController: URL 파싱 후 채용공고 내용 조회 기능 구현
- RecruitService: 입력되는 id값 null 검증
- RecruitDAO: DB에 접근해 조회결과를 DetailDTO로 반환
- recruitment.xml: join연산을 통해 상세정보 조회

Issue: #75

kihong1602 committed:
Feat: 채용지원 기능 구현

- apply.js: 비동기 통신으로 채용지원 요청
- RecruitController: JSON 파싱과 AuthCookie 조회 후 채용공고 지원, 응답결과 JSON타입으로 전송하는 기능 구현
- ApplyInfoDTO: JSON key value 매핑을 위한 DTO 구현
- ApplyService: aptId 생성을 위해 memberId로 연계된 member table id 조회, aptId 생성 기능 구현, DAO 반환 값 검증
- ApplyDAO: DB에 응시지원 데이터를 저장하는 기능 구현
- CookieManager: 쿠키 조회기능 구현
- AptIdFactory: aptId 생성기능 구현

Issue: #76

kihong1602 committed:
Chore: 채용지원 기능 코드 수정

- 채용지원관련 클래스를 채용공고 클래스에 병합

kihong1602 committed:
Chore: mail api, Mybatis 의존성 추가

- Mybatis: 간헐적 라이브러리 로드 오류로 인해 pom.xml로 의존성 추가
- mail api 사용을 위한 jakarta mail 의존성 추가

kihong1602 committed:
Feat: 이메일 발송 기능 구현

- 회원가입 후 이메일인증을 하기위한 이메일 발송기능 구현
- 이메일 인증 페이지 구성
- 기능추가로 인한 기존 Member 관련 클래스 수정

Issue: #78

kihong1602 committed:
Feat: 이메일 인증 기능 구현

- 수신된 이메일에 첨부된 링크 클릭시 인증 프로세스가 실행되도록 구현
- email-auth.jsp 커밋 누락으로 추가

Issue: #79

kihong1602 committed:
Feat: 시험응시권한 인증페이지 구현

- 시험응시를 위한 프론트 페이지 구성
- examId를 이용해 채용공고 정보를 view로 전달

Issue: #84

kihong1602 committed:
Chore: 시험응시권한 인증페이지 수정

- 시험응시를 위한 프론트 페이지 구성
- 구조 수정으로 인한 클래스 삭제 및 수정
- 수험번호를 입력한 후 이메일 발송 으로 기능 변경
- 시험 응시페이지 관련 클래스, View

Issue: #84

kihong1602 committed:
Chore: gitignore 재적용

- gitIgnore 재적용

kihong1602 committed:
Feat: 시험 응시권한 인증 기능 구현

- ExamAuthController: 비동기 통신으로 aptId를 비교하는 기능 구현

Issue: #85

kihong1602 committed:
Feat: 시험 응시페이지 조회

- GET 요청시 시험문제를 조회해 View에 전달
- POST 요청시 작성한 답안을 DB에 저장

Issue: #86

kihong1602 committed:
Chore: 클래스명, 메서드명 변경

- 부정확한 네이밍으로 혼동 가능성이 높아 변경하고, 그에 따라 코드 일부 수정되었습니다.

hossi97 committed:
Merge branch 'main' into main

incutv commented on kihong1602/recrute/src/main/java/com/blanc/recrute/exam/controller/ExamAuthController.java:
```diff
@@ -0,0 +1,85 @@
+
+  private static final Gson GSON = new Gson();
+  private static final String EXAM_AUTH = "examAuth";
+  private static final int HOUR = 60 * 60;
```
적절히 상수로 추출하여 가독성을 높여준 부분 GOOD

incutv commented on kihong1602/recrute/src/main/java/com/blanc/recrute/member/controller/EmailConfirmController.java:
```diff
@@ -0,0 +1,78 @@
+        String authKey = String.valueOf(UUID.randomUUID());
+
+        //이메일 발송
+        boolean check = SendEmailService.mailSend("gmail", "team.recrute1602@gmail.com", "qlhg mjfn oscq fbzo", memberDTO.getEmail(), authKey);
```
이메일 관련 필드들을 DTO로 묶어서 사용했으면 더 좋았을거같아요.

kihong1602 committed:
Chore: DB연결 스레드 연결방식 변경

- ThreadLocal 사용으로 스레드간 독립적인 SqlSession을 유지하게 변경

kihong1602 committed:
Chore: 코드 수정

- Authenticater: 쿠키 null 검증 추가
- CookieManager: 쿠키 null 검증 추가
- DAO: try-catch-finally로 sqlSession commit, rollback, close 하게 변경
- Service: 하드코딩된 응답문자열 상수화
- URLParser: 메서드 name 변경
- ViewResolver: path설정과 forward를 해당 클래스에서 모두 진행하도록 수정
- signin.js: 회원가입기능 비동기처리
- exam*.js: req url 수정

kihong1602 committed:
Chore: 코드 수정 및 Mapper 인터페이스 추가

- 오타방지를 위해 Mybatis Mapper 인터페이스를 추가
- 가독성 향상을 위한 쿼리 id 수정

kihong1602 committed:
Chore: 코드 수정

Controller 클래스들의 불필요한 static 키워드 제거, ViewResolver 클래스 코드변경으로 인한 메서드 수정

kihong1602 committed:
Chore: 유틸 클래스 Word추가, Service 클래스 수정

Service 클래스의 불필요한 static 키워드 제거
Word 클래스를 생성해 자주사용되는 문자열 상수화

kihong1602 committed:
Chore: DAO 클래스 코드수정, Mybatis 연결 관련 수정

DB connection 관련 SqlSession, Mapper 인스턴스 static 해제, 지역변수화
그에 따른 DAO 클래스의 불필요한 static 키워드 제거, 코드 수정

kihong1602 committed:
Feat: 회원가입 요청수단 변경

회원가입 요청을 Form -> Async JSON으로 변경
JSON 데이터를 받을 MemberInfoDTO 생성, 프로세스 성공시 GET 요청을 받을 Controller 추가

kihong1602 committed:
Merge branch 'feature/member', remote-tracking branch 'origin' into feature/exam

kihong1602 committed:
Feat: 시험응시자 이메일 대량발송 기능 구현

exam-email.jsp, exam-email.js: 버튼 클릭시 Async POST 요청
ExamEmailController: 요청을 받아 Service 클래스를 호출
ExamService: DAO에서 email을 List에 담아서 반환받아 이메일 대량발송
ExamDAO: applicant와 member를 Join해 지원자의 이메일을 조회 후 반환

hossi97 commented:
충돌이 많아서 Merge를 할 수가 없는 상황이에요
개인 PC의 "로컬 저장소"에 "팀 저장소의 main브랜치"를 `pull`해서 충돌 수정하고, 
본인의 "개인 저장소"로 다시 `push`해주셔야 할 것 같아요..ㅠㅠ

kihong1602 committed:
Fix: pull and merge
